manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,831 @@
+		for _, config := range local.nw.IPAM.Configs {
+			// the last param of RequestPool is "v6", meaning IPv6, which we
+			// don't support, hence passing "false"
+			poolID, poolIP, _, err := ipam.RequestPool(addressSpace, config.Subnet, config.Range, ipamOpts, false)
+			if err != nil {
+				// typically, if there was an error in requesting a pool, we
+				// would release the pools we've already allocated. However, if
+				// there is an error at this stage, it means the whole object
+				// store is in a bad state, so we don't do that, we just
+				// abandon everything.
+				return errors.ErrBadState(
+					"error reserving ipam pool for network %v: %v",
+					nw.ID, err,
+				)
+			}
+			// each IPAM pool we use has a separate ID referring to it. We also
+			// keep a map of each IP address we have allocated for this network
+			// and what pool it belongs to, so we can deallocate an address
+			// from the correct pool
+			local.pools[poolIP.String()] = poolID
+			// now we just need to reserve the gateway address for this pool.
+			// set the IPAM request address type to "Gateway" to tell the IPAM
+			// driver that's the kind of address we're requesting
+			ipamOpts[ipamapi.RequestAddressType] = netlabel.Gateway
+			// and now, if we have a Gateway address for this network, we need
+			// to allocate it. This condition should never happen; a network
+			// with a valid IPAM config but an empty gateway is a recipe for
+			// disaster.  This is just included for completeness if an older
+			// version has incorrect state
+			if config.Gateway != "" {
+				_, _, err := ipam.RequestAddress(poolID, net.ParseIP(config.Gateway), ipamOpts)
+				if err != nil {
+					return errors.ErrBadState(
+						"error requesting already assigned gateway address %v: %v",
+						err,
+					)
+				}
+				// NOTE(dperny): this check was originally here:
+				// if gwIP.IP.String() != config.Gateway {
+				//   // if we get an IP address from this that isn't the one we
+				//   // requested, that's Very Bad.
+				//   return ErrBadState{
+				//     local.nw.ID,
+				//     fmt.Sprintf("got back gateway ip %v, but requested ip %v", gwIP, config.Gateway),
+				//   }
+				// }
+				// it has been removed because we don't need it. this is a
+				// check that IPAM is behaving correctly. We don't need to
+				// check that IPAM is behaving correctly. Doing so makes this
+				// code less clear and harder to test. This has been left in
+				// so that this bolt of wisdom is shared with future
+				// contributors. A similar check was found in restoreAddress
+				// in the analogous location.
+
+				// most addresses need to be added to the map of
+				// address -> poolid, but we don't need to do this with the
+				// gateway address because it's store in the ipam config with
+				// the subnet, which is the key for the pools map.
+			}
+			// delete the Gateway option from the IPAM options when we're done
+			delete(ipamOpts, ipamapi.RequestAddressType)
```

Comment:
We are mutating the IPAM options map here - is it guaranteed that there was no previous value for `ipamapi.RequestAddressType` in the map?  Otherwise editing and deleting it would be mutating the map, and I wasn't sure if it ever gets written back to the store.

Reply:
I'm 80% sure that this map entry should never be present usually... but 80% isn't really good enough, especially when the API here is so ill-defined, so I've changed the code to save the previous entry if one exists and restore it if one exists (and delete it if one doesn't).